### ISSUE TITLE
Support variables for projections

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12600]
### Tell us your idea!

Now its gets totally wild ;)

In my todays cache GCKGVY a projection is needed to get from last stage to the final:
`Den Cache findest du: (I) (D) (A) Meter in (E) (H) (A) Grad`

So this feature request is to also support variables in bearing/distance for projected coordinates as a waypoint.
For sure nothing important and not to be considered mandatory for the first go live.

### What solution would you suggest?

Not sure about how it could be implemented UI wise. Probably it would need (besides the coords theirselves) additional fields for bearing and distance which support variables/formulas. 
The projection then needs to be calculated into the resulting coords as soon as the variables have valid values.

### Is there an alternative?

No.
For that particular cache I created another waypoint (Final) with same coordinate formula as for the stage upfront (stage 7) and pasted the text for the projection as a note. Still I needed to manually look up the variable values and fill them in to create the correct final waypoint.

**Comment by Volthai7us** [comment]
Hi, I want to work on this issue but I need some guidance. Could you help to figure out where in the code I should go and make changes? This is my first time contributing to an open source project.

**Comment by eddiemuc** [comment]
I don't consider this request to be wild at all. It would be a perfect use case for the new global variables.

That said, I don't think we should implement this right away but first think about how we want to handle Waypoint projections in general. 

As of now, waypoint projection works in a one-time-calculation way: user enters angle+distance into a waypoint, the waypoint is calculated anew and the projected calculation is saved. The projection parameters however are NOT saved.

Introducing variables into this scenario could be as simple as allowing to enter a formula into the angle+distance field. This is easy to implement. However, this would mean that calculation of angle+distance is only done one time from the vars (when they are entered) and projection is immediately calculated. The angle/distance formulas are not saved, and if the values of the vars change at a later time, the projection would not be recalculated.

I would argue that, before introducing vars into this, we should change the way we handle projections: a projection should be specified by entering a reference waypoint, an angle and a distance. Then this waypoint would recalculate itself every time the reference waypoint or angle/distance is changed. Then introducing variabes in angle/distance also makes sense.

However, this would be a bigger task, so I would like to put this up for discussion first.

**Comment by eddiemuc** [comment]
> Hi, I want to work on this issue but I need some guidance. Could you help to figure out where in the code I should go and make changes? This is my first time contributing to an open source project.

Hello @Volthai7us thank you for your cool offer, it is always great to get support for c:geo implementation. However, you picked a rather complicated issue as a starter - it is related to the new global variable feature which is itself still in a very early stage and not even visible in the beta yet. If you're interested in general to start developing for c:geo, maybe it would be better to pick another issue first. E.g. something from the "Good first issue list" (https://github.com/cgeo/cgeo/labels/Good%20first%20issue) 

Or are you specifically interested in the projection / variable topic?

**Comment by Volthai7us** [comment]
Hello @eddiemuc I have to solve a issue from github as a school project this term. And I liked your project. I would like to contribute. I will have free time to studying. Could you offer me any issue I can work on?

**Comment by eddiemuc** [comment]
> Hello @eddiemuc I have to solve a issue from github as a school project this term. And I liked your project. I would like to contribute. I will have free time to studying. Could you offer me any issue I can work on?

What about e.g. #11664 as a starter?

However, the first thing before starting development would be to setup your development environment. Check following guides for this:
* The general workflow and prerequisites to work on c:geo: https://github.com/cgeo/cgeo/blob/master/README.md
* A detailled guide to set up all the git/github stuff: https://github.com/cgeo/cgeo/wiki/Working-on-c%3Ageo-for-git-beginners

**Comment by andrcoding** [comment]
Not directly linked to waypoint projection and variables but it came into my mind reading this issue: Sometimes I have the issue that I have to count some objects which are not quite clear resulting in having issues finding the final cache (The number is used in the final gc equation). Therefore, it could be helpful to select one variable and define some values/a range for that variable to get some possible locations shown on a map. Usually, after that it is clear, which solution is correct.

**Comment by moving-bits** [comment]
@andrcoding 
See #12588 for an approach to this.

**Comment by eddiemuc** [comment]
Ok, before we talk about using vars in projections (which actually would be quite easy to achieve) we have to talk about how we want projections to be handled overall in c:geo.

Today's handling is messy IMHO. There are fields angle/distance in a waypoint which kinda looks like as if they were persisted on "save waypoint" - but in fact those are just transient fields whose value will be lost on "save", as well as the original waypoint coordinate. The waypoints coordinate is simply recalculated on save and the result stored in the waypoint coord. Angle, distance as well as the original waypoint's coordinate are lost after that. I have constant problems with this behavior out in the field, leading to manual creation of "original" waypoints and "projected" waypoints in order not to loose my data. Needless to say, this behaviour also makes it impossible to let c:geo recalculate projections if any of the "source values" (original coord, angle, distance) changes.

I have two principal ideas to approach this in a better way:

**Idea 1: make "projection" always reference another waypoint as its "source coordinate"**


Waypoints can either be of type "lat/lon" (either calculated or constant, doesn't matter) or "projections". "Projection" waypoints have a "source" waypoint which they reference. Projection angle and distance are stored on the "projection" waypoint. The "projection" waypoints coordinate is calculated with these values. Whenever the coord in the "source waypoint" changes, the "projection waypoint" is recalculated as well automatically.

(This is the way GCBuddy handles this case. GCBuddy also supports a third time "offset", which also takes a source waypoint but shifts it by a user-given lat/lon-offset)

Following is a mock-up on how such a projected waypoint could look like:

![image](https://user-images.githubusercontent.com/6909759/154857914-49fcef51-89f0-43ad-a35f-b912e2639d45.png)


**Idea 2: keep both "source coord" as well as "projected coord" in one waypoint**

Waypoints have a coordinate as today. If a projection is given however, they will calculate a second (projected) coordinate. This second coordinate is the shown and used waypoint coordinate. However, in the waypoint edit dialog one can see both source as well as projected coordinate.

(This way would be kinda closer to the way projection is handled in c:geo today, but of course two coords have to be kept inside the waypoint in this case)

Following is a mock-up how idea 2 could look in waypoint edit dialog:

![image](https://user-images.githubusercontent.com/6909759/154857569-5da05087-0836-4fba-a6d0-709fab3b4ebd.png)


I would be interested in opinions and maybe further ideas about this.

**Comment by eddiemuc** [comment]
Hm I guess it might make sense to start with idea 2 since it is closer to what users are used to and easier to implement.

Also it can be extended to idea 1 later. And idea 1 will get complex again. Among other things cyclic dependency detection between waypoints needs detection.

But idea 1 would be so much cooler imho...

**Comment by Lineflyer** [comment]
Need to read through this issue next days to provide feedback.
Sorry...to much going on here and also still work to do ;)

**Comment by moving-bits** [comment]
Some spontaneous thought on this:
- Both variants give making a projection much visual attention, although it's just one of several things to do with a waypoint.
- Option 1 would introduce a relationship to sort of a "parent" waypoint. Something which would be useful for generated waypoints as well.
- Option 1: What shall be done if a "parent" waypoint changes its coordinates - will a fresh projection be performed automatically?
- Option 2 seems to be sort of a logical workflow - you have source coordinates, do something to them, and see a result. - Very welcomed to see both source and target coords.
- How to handle multiple projections? Have source coords, do a projection, and the do a second projection from there. (Probably need to create fresh waypoints for it. But they can't be a simple copy then, they need to use the target coordinates as new source coordinates for this use case to work. How to differentiate?) - But this may be a rare edge case.
- Not directly related: Can we switch the distance and the angle fields? 99% of caches with projections list distance first, and angle second.
- Maybe we can put distance and angle fields in the same line, there should be enough available space. This would make the whole thing a bit more compact.

**Comment by Lineflyer** [comment]
> Not directly related: Can we switch the distance and the angle fields? 99% of caches with projections list distance first, and angle second.

That deserves a quick reaction, before following up in more detail:
Oh yes..that happens to me also all the time.

**Comment by eddiemuc** [comment]
Feedback from me to the very valid questions of @moving-bits.

In long term I would be a fan of option 2, but short-term option 1 is much easier to implement and is closer to what users are used to in cgeo.
 
> * Both variants give making a projection much visual attention, although it's just one of several things to do with a waypoint.

Agree, this is not wanted. IMHO projection already today has too much visual attention (with the two fields always present). I woudl actually introduce some kind of collapse/expand solution for this. Maybe a dropdown where one can select "Lat/Lon (no waypoint calc)", "projection" and "offset", and necessary input fields only become visible for selected function. And if we have more ideas for waypoint calcs later they can be added. GCBuddy has it this way, you may take a look at it.

> * Option 1 would introduce a relationship to sort of a "parent" waypoint. Something which would be useful for generated waypoints as well.

Correct

> * Option 1: What shall be done if a "parent" waypoint changes its coordinates - will a fresh projection be performed automatically?

I would assume automatic recalculation. Same as when variable values change in calculacted coordinates (or, if this is implemented, in formula expressions used for ange/distance in projection)

> * Option 2 seems to be sort of a logical workflow - you have source coordinates, do something to them, and see a result. - Very welcomed to see both source and target coords.

Yes, but we have to be very disciplined NOT TO USE the source column anywhere outside this waypoint! (and in the end having to deal with waypoints having up to two different coordinates in every part of c:geo...)

> * How to handle multiple projections? Have source coords, do a projection, and the do a second projection from there. (Probably need to create fresh waypoints for it. But they can't be a simple copy then, they need to use the target coordinates as new source coordinates for this use case to work. How to differentiate?) - But this may be a rare edge case.

See GCBuddy for a working solution. Basically, it is allowed to create "chains" and "trees" (but we need a "cycle" detection to prevent using WP1 as base for WP2, WP2 as base for WP3 and WP3 as base for WP1...)

**Comment by eddiemuc** [comment]
BTW the "chain"/"Tree"-but-not-"cycle" discussion is already present (and implemented) with the cache variables of course.

**Comment by eddiemuc** [comment]
> > * How to handle multiple projections? Have source coords, do a projection, and the do a second projection from there. (Probably need to create fresh waypoints for it. But they can't be a simple copy then, they need to use the target coordinates as new source coordinates for this use case to work. How to differentiate?) - But this may be a rare edge case.
> 
> See GCBuddy for a working solution. Basically, it is allowed to create "chains" and "trees" (but we need a "cycle" detection to prevent using WP1 as base for WP2, WP2 as base for WP3 and WP3 as base for WP1...)

...but this is only possible with idea 2 of course. With idea 1 this is not possible (only copying waypoints as initial start for next projection; but later changes to the copied "source waypoint" won't change the "target waypoint" then)

**Comment by fm-sys** [comment]
I recently had a cache `Peile F*H*H*H-9m in Richtung F*F-H*H*H*H-9°`.
 
However, I had a problem: I was unsure about the value for one variable. 
So if the waypoint projection gets refactored, I would really like to see the possibility to generate multiple WPs via range variables (like it is possible in normal formulas).
